### PR TITLE
feat(lifecycle): conditionally enable partitions

### LINKED ISSUE #4463: Enable craft-parts partitions feature
### What needs to get done

Snapcraft should enable the partitions feature, define the `default` partition, and use the `default` partition for snap data.  There should be no change in behavior for users.

New spread tests should be added to test the partitions feature. For example:
- the envvars `$CRAFT_STAGE`, `$CRAFT_DEFAULT_STAGE`, `$CRAFT_PRIME`, and `$CRAFT_DEFAULT_PRIME` work as expected in parts
- keywords like `stage`, `prime`, and `organize` work with partitions

### Why it needs to get done

To support components in snapcraft

### LINKED ISSUE #4465: Create partitions for components
### What needs to get done

In `snapcraft/parts/lifecycle.py`, snapcraft should:
- get the list of components from the `Project` model
- create a list of partitions
- pass the partitions to the `LifecycleManager`

A new spread test should verify that partitions exist in the file tree and files can be organized into a component’s partition.

### Why it needs to get done

To support components in snapcraft

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

- Conditionally enable partitions if components are defined and base is `devel`, `core24`, or newer
- Create partitions from components
- Create envvars for partitions (e.g. `SNAPCRAFT_COMPONENT_<name>_STAGE`)

TODO:
- [x] improve spread test to the partitions are created and `stage`, `prime`, and `organize` keywords work as-expected

Replaces #4487

Fixes #4463 
Fixes #4465 

(CRAFT-2307)
(CRAFT-2311)

### CODE REVIEW

File: snapcraft/parts/lifecycle.py [hunk 1]
```diff
@@ -90,6 +92,8 @@ def run(command_name: str, parsed_args: "argparse.Namespace") -> None:
 
     build_count = utils.get_parallel_build_count()
 
+    partitions = _validate_and_get_partitions(yaml_data)
```
Comment (mr-cal) [inline snapcraft/parts/lifecycle.py:95]: Try unmarshalling just the components keyword (and base, type, build-base possibly)

Comment (mr-cal) [reply snapcraft/parts/lifecycle.py:95]: allow for core22 to use partitions

**Comment by mr-cal** [comment]
Spread test failures look like a mistake on my end (something with the project or metadata model).  Looking into now

**Comment by mr-cal** [comment]
> Spread test failures look like a mistake on my end (something with the project or metadata model). Looking into now

Nevermind, I forgot to push my changes...

File: snapcraft/projects.py [hunk 2]
```diff
@@ -884,6 +900,52 @@ def unmarshal(cls, data: Dict[str, Any]) -> "ArchitectureProject":
         return architectures
 
 
+class ComponentProject(ProjectModel, extra=pydantic.Extra.ignore):
```
Comment (cmatsuoka) [inline snapcraft/projects.py:903]: @lengau suggested we can later have support to handle all early project extractions in a single model.

**Review by lengau — APPROVED** [review]
Looks good! I found one (minor UX) bug, but it's not in this code

**Pull request merged**